libraries/c_sdk/standard/https/src/iot_https_client.c

```diff
@@ -0,0 +1,3326 @@
+
+static int _httpParserOnMessageBeginCallback( http_parser * pHttpParser )
+{
+    int retVal = KEEP_PARSING;
+
+    IotLogDebug( "Parser: Start of HTTPS Response message." );
+
+    _httpsResponse_t * pHttpsResponse = ( _httpsResponse_t * ) ( pHttpParser->data );
+    /* Set the state of the parser. The headers are at the start of the message always. */
+    pHttpsResponse->parserState = PARSER_STATE_IN_HEADERS;
+    return retVal;
+}
+
+/*-----------------------------------------------------------*/
+
+static int _httpParserOnStatusCallback( http_parser * pHttpParser,
+                                        const char * pLoc,
+                                        size_t length )
+{
+    _httpsResponse_t * pHttpsResponse = ( _httpsResponse_t * ) ( pHttpParser->data );
+
+    IotLogDebug( "Parser: Status %.*s retrieved from HTTPS response.", length, pLoc );
+
+    /* Save the status code so it can be retrieved with IotHttpsClient_ReadResponseStatus(). */
+    pHttpsResponse->status = ( uint16_t ) ( pHttpParser->status_code );
+
+    /* If we are parsing the network data received in the header buffer then we
+     * increment pHttpsResponse->pHeadersCur. The status line in the response is
+     * part of the data stored in header buffer _httpResponse->pHeaders. */
+    if( pHttpsResponse->bufferProcessingState == PROCESSING_STATE_FILLING_HEADER_BUFFER )
+    {
+        /* pHeadersCur will never exceed the pHeadersEnd here because PROCESSING_STATE_FILLING_HEADER_BUFFER
+         * indicates we are currently in the header buffer and the total size of the header buffer is passed
+         * into http_parser_execute() as the maximum length to parse. */
+        pHttpsResponse->pHeadersCur = ( uint8_t * ) ( pLoc += length );
+    }
+
+    return KEEP_PARSING;
+}
+
+/*-----------------------------------------------------------*/
+
+static int _httpParserOnHeaderFieldCallback( http_parser * pHttpParser,
+                                             const char * pLoc,
+                                             size_t length )
+{
+    IotLogDebug( "Parser: HTTPS header field parsed %.*s", length, pLoc );
+
+    _httpsResponse_t * pHttpsResponse = ( _httpsResponse_t * ) ( pHttpParser->data );
+
+    /* If we are parsing the network data received in the header buffer then we can increment
+     * pHttpsResponse->pHeadersCur. */
+    if( pHttpsResponse->bufferProcessingState == PROCESSING_STATE_FILLING_HEADER_BUFFER )
+    {
+        pHttpsResponse->pHeadersCur = ( uint8_t * ) ( pLoc += length );
+    }
+
+    /* If the IotHttpsClient_ReadHeader() was called, then we check for the header field of interest. */
+    if( pHttpsResponse->bufferProcessingState == PROCESSING_STATE_SEARCHING_HEADER_BUFFER )
+    {
+        if( pHttpsResponse->readHeaderFieldLength != length )
+        {
+            pHttpsResponse->foundHeaderField = false;
+        }
+        else if( strncmp( pHttpsResponse->pReadHeaderField, pLoc, length ) == 0 )
```

Comment:
No because we check the lengths matching above:
```
        if( pHttpsResponse->readHeaderFieldLength != length )
        {
            pHttpsResponse->foundHeaderField = false;
        }
```